tensorflow/core/util/gpu_device_functions.h

```diff
@@ -263,34 +284,34 @@ __device__ T CudaShuffleSync(unsigned mask, T value, int src_lane,
 // Variant of the (undocumented) version from the CUDA SDK, but using unsigned
 // instead of float for lo and hi (which is incorrect with ftz, for example).
 // See b/69446944.
-__device__ inline double CudaShuffleSync(unsigned mask, double value,
+__device__ inline double GpuShuffleSync(unsigned mask, double value,
                                          int src_lane, int width = warpSize) {
+  unsigned lo, hi;
 #if GOOGLE_CUDA
-  auto tmp = __double_as_longlong(value);
-  auto lo = static_cast<unsigned>(tmp);
-  auto hi = static_cast<unsigned>(tmp >> 32);
-  hi = CudaShuffleSync(mask, hi, src_lane, width);
-  lo = CudaShuffleSync(mask, lo, src_lane, width);
-  return __longlong_as_double(static_cast<uint64_t>(hi) << 32 | lo);
+  asm volatile("mov.b64 {%0,%1}, %2;" : "=r"(lo), "=r"(hi) : "d"(value));
+  hi = GpuShuffleSync(mask, hi, src_lane, width);
+  lo = GpuShuffleSync(mask, lo, src_lane, width);
+  asm volatile("mov.b64 %0, {%1,%2};" : "=d"(value) : "r"(lo), "r"(hi));
+  return value;
 #elif TENSORFLOW_USE_ROCM
-  auto tmp = static_cast<uint64_t>(value);
-  auto lo = static_cast<unsigned>(tmp);
-  auto hi = static_cast<unsigned>(tmp >> 32);
+  uint64_t tmp = static_cast<uint64_t>(value);
```

Comment:
@jerryyin to keep the PR simple, could you not change logic for these shuffle routines?

Reply:
Thanks for reminding. The file version appeared here is before the weekly upstream sync, making it out of date. Checking out the current develop-upstream version gpu_device_functions.h seems to fix it.